File: tests/oracle/action_pool_clone_filter_1204.cc
/*
 * ACTIONPOOL-CLONE-FILTER-0001: locked Ghidra 12.0.4 behavioral oracle.
 *
 * A synthetic registered "universal" ActionPool is derived through the real
 * ActionDatabase::setCurrent -> deriveAction -> ActionPool::clone path.  The
 * observation preserves allrules and perop order and exposes the inherited
 * Action/Rule runtime fields so fresh-clone versus copied-state behavior is
 * explicit.  Test-only access changes visibility only; production bodies are
 * linked unchanged from the locked libdecomp archive.
 */

#include <bits/stdc++.h>
#define class struct
#define private public
#define protected public
#include "action.hh"
#undef class
#undef private
#undef protected

using namespace ghidra;
using std::cout;
using std::runtime_error;

class ScriptRule final : public Rule {
  uint4 constructor_flags;
  string constructor_name;
  vector<uint4> opcodes;

public:
  ScriptRule(const string &group, uint4 flags, const string &name,
             const vector<uint4> &ops)
      : Rule(group, flags, name), constructor_flags(flags),
        constructor_name(name), opcodes(ops) {}

  Rule *clone(const ActionGroupList &grouplist) const override {
    if (!grouplist.contains(getGroup()))
      return (Rule *)0;
    return new ScriptRule(getGroup(), constructor_flags, constructor_name,
                          opcodes);
  }

  void getOpList(vector<uint4> &result) const override {
    result.insert(result.end(), opcodes.begin(), opcodes.end());
  }
};

static string opcode_name(uint4 opcode) {
  if (opcode == CPUI_COPY) return "copy";
  if (opcode == CPUI_INT_ADD) return "int_add";
  if (opcode == CPUI_INT_SUB) return "int_sub";
  return "unknown";
}

static ActionPool *build_source(void) {
  ActionPool *pool = new ActionPool(
      Action::rule_repeatapply | Action::rule_onceperfunc |
          Action::rule_warnings_given,
      "universal");
  pool->addRule(new ScriptRule("alpha", Rule::rule_debug, "a",
                               {CPUI_COPY, CPUI_INT_ADD}));
  pool->addRule(new ScriptRule("beta", 0, "b", {CPUI_INT_ADD}));
  pool->addRule(new ScriptRule("alpha", Rule::warnings_on, "c",
                               {CPUI_COPY}));
  pool->addRule(new ScriptRule("gamma", 0, "d",
                               {CPUI_INT_SUB, CPUI_COPY}));

  pool->status = Action::status_mid;
  pool->breakpoint = Action::break_start | Action::break_action;
  pool->count_tests = 17;
  pool->count_apply = 11;
  for (size_t i = 0; i < pool->allrules.size(); ++i) {
    Rule *rule = pool->allrules[i];
    rule->flags |= Rule::type_disable | Rule::warnings_on |
                   Rule::warnings_given;
    rule->breakpoint = (uint4)(9 + i);
    rule->count_tests = (uint4)(30 + i);
    rule->count_apply = (uint4)(20 + i);
  }
  return pool;
}

static void emit_pool(const string &case_name, Action *action) {
  if (action == (Action *)0) {
    cout << "case=" << case_name << " null=1\n";
    return;
  }
  ActionPool *pool = dynamic_cast<ActionPool *>(action);
  if (pool == (ActionPool *)0)
    throw runtime_error("derived action is not ActionPool");

  cout << "case=" << case_name << " null=0 name=" << pool->getName()
       << " flags=" << pool->flags << " status=" << pool->status
       << " breakpoint=" << pool->breakpoint
       << " tests=" << pool->count_tests << " apply=" << pool->count_apply
       << " rules=" << pool->allrules.size() << '\n';
  for (size_t i = 0; i < pool->allrules.size(); ++i) {
    Rule *rule = pool->allrules[i];
    vector<uint4> ops;
    rule->getOpList(ops);
    cout << "rule=" << case_name << ':' << i << " name=" << rule->getName()
         << " group=" << rule->getGroup() << " flags=" << rule->flags
         << " breakpoint=" << rule->breakpoint
         << " tests=" << rule->count_tests << " apply=" << rule->count_apply
         << " opcodes=";
    for (size_t j = 0; j < ops.size(); ++j) {
      if (j != 0) cout << ',';
      cout << opcode_name(ops[j]);
    }
    cout << '\n';
  }
  const uint4 observed[] = {CPUI_COPY, CPUI_INT_ADD, CPUI_INT_SUB};
  for (uint4 opcode : observed) {
    cout << "perop=" << case_name << ':' << opcode_name(opcode) << " rules=";
    const vector<Rule *> &rules = pool->perop[opcode];
    for (size_t i = 0; i < rules.size(); ++i) {
      if (i != 0) cout << ',';
      cout << rules[i]->getName();
    }
    cout << '\n';
  }
}

static void derive_and_emit(ActionDatabase &database, const string &name) {
  size_t before = database.actionmap.size();
  Action *first = database.setCurrent(name);
  size_t after_first = database.actionmap.size();
  Action *second = database.setCurrent(name);
  size_t after_second = database.actionmap.size();
  bool entry = database.actionmap.find(name) != database.actionmap.end();
  bool cached = first == second && after_first == after_second &&
                after_first == before + 1;
  cout << "derive=" << name << " entry=" << (entry ? 1 : 0)
       << " cached=" << (cached ? 1 : 0)
       << " current=" << database.getCurrentName() << '\n';
  emit_pool(name, first);
}

int main(void) {
  ActionDatabase database;
  ActionPool *source = build_source();
  emit_pool("source", source);
  database.registerAction("universal", source);

  const char *all[] = {"alpha", "beta", "gamma", (const char *)0};
  const char *custom[] = {"alpha", "gamma", (const char *)0};
  const char *empty[] = {(const char *)0};
  database.setGroup("all", all);
  database.setGroup("custom", custom);
  database.setGroup("empty", empty);

  derive_and_emit(database, "all");
  derive_and_emit(database, "custom");
  derive_and_emit(database, "empty");
  return 0;
}

File: tests/oracle/action_pool_clone_filter_1204.rs
//! ACTIONPOOL-CLONE-FILTER-0001 Rust comparand.
//!
//! This drives production ActionDatabase derivation and ActionPool cloning.
//! Only the synthetic Rule supplies its own virtual clone, exactly like a
//! concrete Ghidra Rule; no hand-written expected output is embedded.

use rugra::action::{
    action_flags, status_flags, Action, ActionDatabase, ActionGroupList, ActionPool, ActionState,
    Rule,
};
use rugra::funcdata::Funcdata;
use rugra::op::PcodeOp;
use rugra::opcodes::OpCode;
use rugra::Result;
use std::sync::{Arc, RwLock};

struct ScriptRule {
    group: &'static str,
    constructor_flags: u32,
    name: &'static str,
    opcodes: Vec<OpCode>,
    runtime_flags: u32,
    breakpoint: u32,
    count_tests: u32,
    count_apply: u32,
}

impl ScriptRule {
    fn new(
        group: &'static str,
        constructor_flags: u32,
        name: &'static str,
        opcodes: Vec<OpCode>,
        dirty: bool,
        ordinal: u32,
    ) -> Self {
        Self {
            group,
            constructor_flags,
            name,
            opcodes,
            runtime_flags: if dirty {
                constructor_flags | 1 | 4 | 8
            } else {
                constructor_flags
            },
            breakpoint: if dirty { 9 + ordinal } else { 0 },
            count_tests: if dirty { 30 + ordinal } else { 0 },
            count_apply: if dirty { 20 + ordinal } else { 0 },
        }
    }
}

impl Rule for ScriptRule {
    fn apply_op(&self, _op: &Arc<RwLock<PcodeOp>>, _fd: &mut Funcdata) -> Result<i32> {
        Ok(0)
    }

    fn get_name(&self) -> &str {
        self.name
    }

    fn get_opcodes(&self) -> Vec<OpCode> {
        self.opcodes.clone()
    }

    fn clone_for_groups(&self, grouplist: &ActionGroupList) -> Option<Box<dyn Rule>> {
        if !grouplist.contains(self.group) {
            return None;
        }
        Some(Box::new(Self::new(
            self.group,
            self.constructor_flags,
            self.name,
            self.opcodes.clone(),
            false,
            0,
        )))
    }

    fn get_group(&self) -> &str {
        self.group
    }
    fn get_rule_flags(&self) -> u32 {
        self.runtime_flags
    }
    fn get_breakpoint(&self) -> u32 {
        self.breakpoint
    }
    fn get_num_tests(&self) -> u32 {
        self.count_tests
    }
    fn get_num_apply(&self) -> u32 {
        self.count_apply
    }
}

fn opcode_name(opcode: OpCode) -> &'static str {
    match opcode {
        OpCode::CPUI_COPY => "copy",
        OpCode::CPUI_INT_ADD => "int_add",
        OpCode::CPUI_INT_SUB => "int_sub",
        _ => "unknown",
    }
}

fn build_source() -> ActionPool {
    let mut pool = ActionPool::with_flags(
        "universal",
        action_flags::RULE_REPEATAPPLY
            | action_flags::RULE_ONCEPERFUNC
            | action_flags::RULE_WARNINGS_GIVEN,
    );
    pool.add_rule(Box::new(ScriptRule::new(
        "alpha",
        2,
        "a",
        vec![OpCode::CPUI_COPY, OpCode::CPUI_INT_ADD],
        true,
        0,
    )));
    pool.add_rule(Box::new(ScriptRule::new(
        "beta",
        0,
        "b",
        vec![OpCode::CPUI_INT_ADD],
        true,
        1,
    )));
    pool.add_rule(Box::new(ScriptRule::new(
        "alpha",
        4,
        "c",
        vec![OpCode::CPUI_COPY],
        true,
        2,
    )));
    pool.add_rule(Box::new(ScriptRule::new(
        "gamma",
        0,
        "d",
        vec![OpCode::CPUI_INT_SUB, OpCode::CPUI_COPY],
        true,
        3,
    )));
    pool
}

fn emit_pool(case_name: &str, action: Option<&dyn Action>, state: Option<&ActionState>) {
    let Some(pool) = action.and_then(Action::as_action_pool) else {
        println!("case={case_name} null=1");
        return;
    };
    let state = state.expect("non-null pool state");
    println!(
        "case={case_name} null=0 name={} flags={} status={} breakpoint={} tests={} apply={} rules={}",
        pool.get_name(),
        state.flags,
        state.status,
        state.breakpoint,
        state.count_tests,
        state.count_apply,
        pool.rules().len(),
    );
    for (index, rule) in pool.rules().iter().enumerate() {
        let opcodes = rule
            .get_opcodes()
            .into_iter()
            .map(opcode_name)
            .collect::<Vec<_>>()
            .join(",");
        println!(
            "rule={case_name}:{index} name={} group={} flags={} breakpoint={} tests={} apply={} opcodes={opcodes}",
            rule.get_name(),
            pool.rule_group(index),
            rule.get_rule_flags(),
            rule.get_breakpoint(),
            rule.get_num_tests(),
            rule.get_num_apply(),
        );
    }
    for opcode in [
        OpCode::CPUI_COPY,
        OpCode::CPUI_INT_ADD,
        OpCode::CPUI_INT_SUB,
    ] {
        println!(
            "perop={case_name}:{} rules={}",
            opcode_name(opcode),
            pool.rule_names_for_opcode(opcode).join(","),
        );
    }
}

fn derive_and_emit(database: &mut ActionDatabase, name: &str) {
    let was_absent = !database.has_action_entry(name);
    database.set_current(name);
    let first = database
        .get_action(name)
        .map(|action| action as *const dyn Action as *const ());
    database.set_current(name);
    let second = database
        .get_action(name)
        .map(|action| action as *const dyn Action as *const ());
    let entry = database.has_action_entry(name);
    let cached = was_absent && entry && first == second;
    println!(
        "derive={name} entry={} cached={} current={}",
        u8::from(entry),
        u8::from(cached),
        database.get_current_name(),
    );
    let action = database.get_action(name);
    let state = action.map(|action| ActionState::new(action.get_flags()));
    emit_pool(name, action, state.as_ref());
}

fn main() {
    let source = build_source();
    let mut source_state = ActionState::new(source.get_flags());
    source_state.status = status_flags::STATUS_MID;
    source_state.breakpoint = 5;
    source_state.count_tests = 17;
    source_state.count_apply = 11;
    emit_pool("source", Some(&source), Some(&source_state));

    let mut database = ActionDatabase::new();
    database.register_action(Box::new(source));
    database.set_group("all", &["alpha", "beta", "gamma"]);
    database.set_group("custom", &["alpha", "gamma"]);
    database.set_group("empty", &[]);

    derive_and_emit(&mut database, "all");
    derive_and_emit(&mut database, "custom");
    derive_and_emit(&mut database, "empty");
}
